exp/runtime/internal/controllers/warmup.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+)
+
+const (
+	defaultWarmupTimeout  = 60 * time.Second
+	defaultWarmupInterval = 2 * time.Second
+)
+
+var _ manager.LeaderElectionRunnable = &warmupRunnable{}
+
+// warmupRunnable is a controller runtime LeaderElectionRunnable. It warms up the registry on controller start.
+type warmupRunnable struct {
+	Client         client.Client
+	APIReader      client.Reader
+	RuntimeClient  runtimeclient.Client
+	warmupTimeout  time.Duration
+	warmupInterval time.Duration
+}
+
+// NeedLeaderElection satisfies the controller runtime LeaderElectionRunnable interface.
```

Comment:
nit: can we add a note explaining why are we returning true (to ensure the operation happens before the any controller start)